app/models/assignment.rb

```diff
@@ -13,23 +13,38 @@ class Assignment < ActiveRecord::Base
   validates :mixpanel_result, inclusion: { in: %w(success failure) }, allow_nil: true
   validate :variant_must_exist
 
-  scope :unsynced_to_mixpanel, -> { where("mixpanel_result = 'failure' OR mixpanel_result IS NULL") }
-
   normalize_attributes :mixpanel_result
 
   delegate :name, to: :split, prefix: true
 
-  class << self
-    def to_hash
-      Hash[all.includes(:split).map { |a| [a.split.name.to_sym, a.variant.to_sym] }]
-    end
+  scope :for_presentation, ->(app_build: nil) do
+    q = excluding_decision_overrides
+    app_build.present? ? q.for_app_build(app_build) : q.for_active_splits
+  end
 
-    def for_presentation(built_at: nil)
-      q = joins(:split)
-        .where("splits.decided_at is null or assignments.updated_at > splits.decided_at")
-      q = built_at.nil? ? q.where("splits.finished_at is null") : q.where("splits.finished_at > ?", built_at)
-      q
-    end
+  scope :for_app_build, ->(app_build) do
+    for_active_splits(as_of: app_build.built_at)
+      .excluding_incomplete_features(app_id: app_build.app_id, version: app_build.version)
+  end
+
+  scope :excluding_decision_overrides, -> do
+    joins(:split).where('splits.decided_at is null or assignments.updated_at > splits.decided_at')
+  end
+
+  scope :for_active_splits, ->(as_of: nil) do
+    joins(:split).merge(Split.active(as_of: as_of))
+  end
+
+  scope :excluding_incomplete_features, ->(app_id:, version:) do
+    joins(:split).where(<<~SQL, app_id, AppVersion.new(version).to_pg_array)
```

Comment:
In theory this isn't actually assignment-specific, so you could move it up to `Split` alongside `active`, but at that point I guess it's just kind of spreading out the concepts:

```ruby
joins(:split).merge(Split.for_complete_features(app_id: app_id, version: version))
```

So happy leaving this as-is.

Reply:
I may refactor that direction when making split weighting registries behave, or maybe that’ll work better in ruby, tbd.